flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala

```diff
@@ -1463,21 +1465,23 @@ class CodeGenerator(
     */
   def addReusableFunction(function: UserDefinedFunction): String = {
     val classQualifier = function.getClass.getCanonicalName
-    val fieldTerm = s"function_${classQualifier.replace('.', '$')}"
+    val functionSerializedData = serialize(function)
+    val fieldTerm =
+      s"""
+         |function_${classQualifier.replace('.', '$')}_${DigestUtils.md5Hex(functionSerializedData)}
```

Comment:
>But from tableAPI to SqlFunction the creation process has some work to do because the current UDF parameters do not take effect.

Do you mean "from Java Table API to SqlFunction" ?  If yes, the Java Table API (i.e. parse expression from a string) will instantiate a new instance of ScalarFunction when lookup this function call (the line is https://github.com/apache/flink/blob/master/flink-libraries/flink-table/src/main/scala/org/apache/flink/table/validate/FunctionCatalog.scala#L84). TableFunction will reuse the object created by user. You can follow the way of TableFunction to fix this.

BTW, scalar UDF in Scala Table API works good, right?